client.go

```diff
@@ -2012,6 +2029,149 @@ func (c *Client) MigrateFrom(name string, operation string, certificate string,
+		// Connect to source server websockets.
+		sourceControlConn, err := sourceClient.Websocket(sourceOperation, sourceControlSecret)
+		if err != nil {
+			return nil, err
+		}
+		sourceFsConn, err := sourceClient.Websocket(sourceOperation, sourceFsSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		var sourceCriuConn *websocket.Conn
+		if criuSecret {
+			sourceCriuConn, err = sourceClient.Websocket(sourceOperation, sourceCriuSecret)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// Post to target server and request and retrieve a set of
+		// websockets + secrets matching those of the source server.
+		resp, err := c.post("containers", body, Async)
+		if err != nil {
+			return nil, err
+		}
+
+		destSecrets := map[string]string{}
+		op, err := resp.MetadataAsOperation()
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range *op.Metadata {
+			destSecrets[k] = v.(string)
+		}
+
+		destControlSecret, ok := destSecrets["control"]
+		if !ok {
+			return nil, fmt.Errorf("Missing control secret")
+		}
+		destFsSecret, ok := destSecrets["fs"]
+		if !ok {
+			return nil, fmt.Errorf("Missing fs secret")
+		}
+		destCriuSecret, ok := destSecrets["criu"]
+		if criuSecret && !ok || !criuSecret && ok {
+			return nil, fmt.Errorf("Missing criu secret")
+		}
+
+		// Connect to dest server websockets.
+		destControlConn, err := c.Websocket(resp.Operation, destControlSecret)
+		if err != nil {
+			return nil, err
+		}
+		destFsConn, err := c.Websocket(resp.Operation, destFsSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		var destCriuConn *websocket.Conn
+		if criuSecret {
+			destCriuConn, err = c.Websocket(resp.Operation, destCriuSecret)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// Let client shovel data from src to dest server.
+		capacity := 4
+		if criuSecret {
+			capacity += 2
+		}
+		syncChan := make(chan error, capacity)
+		defer close(syncChan)
+
+		proxy := func(src *websocket.Conn, dest *websocket.Conn) {
+			for {
+				mt, r, err := src.NextReader()
+				if err != nil {
+					if err != io.EOF {
+						syncChan <- err
+						break
+					}
+				}
+
+				/* The gorilla websockets framework doesn't
+				*  allow concurrent writes. */
+				var controlLock sync.RWMutex
+				controlLock.Lock()
+				w, err := dest.NextWriter(mt)
+				if err != nil {
+					controlLock.Unlock()
+					syncChan <- err
+					break
+				}
+
+				if _, err := io.Copy(w, r); err != nil {
+					controlLock.Unlock()
+					syncChan <- err
+					break
+				}
+
+				if err := w.Close(); err != nil {
+					controlLock.Unlock()
+					syncChan <- err
+					break
+				}
+
+				controlLock.Unlock()
+			}
+		}
+
+		go proxy(sourceControlConn, destControlConn)
+		go proxy(destControlConn, sourceControlConn)
+		go proxy(sourceFsConn, destFsConn)
+		go proxy(destFsConn, sourceFsConn)
+		if criuSecret {
+			go proxy(sourceCriuConn, destCriuConn)
+			go proxy(destCriuConn, sourceCriuConn)
+		}
+
+		for i := 0; i < cap(syncChan); i++ {
+			<-syncChan
```

Comment:
I think we should not log the errors since we only close the websocket but do not write an actual close message to it. That means for the filesystem websocket we would always report errors.